File: src/refresh.ts
import { Credentials, CredentialsData } from './credentials';
import { discoverOIDC, resolveStarkDomain, OIDCDeps } from './oidc';

export interface RefreshDeps {
  fetch?: (url: string, init?: any) => Promise<{ ok: boolean; status: number; json: () => Promise<any> }>;
  credentials?: Credentials;
}

export interface RefreshResult {
  success: boolean;
  credentials?: CredentialsData;
  error?: string;
}

const CLIENT_ID = 'monocle-cli';
const REFRESH_TOKEN_TTL_DAYS = 30;

/**
 * Refresh the access token using the refresh_token grant
 */
export async function refreshAccessToken(
  currentCredentials: CredentialsData,
  deps?: RefreshDeps
): Promise<RefreshResult> {
  const fetchFn = deps?.fetch ?? globalThis.fetch;
  const credentials = deps?.credentials ?? new Credentials();

  // Discover token endpoint and router URL
  let tokenEndpoint: string;
  let discoveredRouterUrl: string | undefined;
  try {
    const starkDomain = resolveStarkDomain(currentCredentials.tenant_domain);
    const oidc = await discoverOIDC(starkDomain, { fetch: deps?.fetch } as OIDCDeps);
    tokenEndpoint = oidc.token_endpoint;
    discoveredRouterUrl = oidc.router_url;
  } catch (err: any) {
    return { success: false, error: `OIDC Discovery failed: ${err.message}` };
  }

  // Request new tokens
  const body = new URLSearchParams({
    grant_type: 'refresh_token',
    refresh_token: currentCredentials.refresh_token,
    client_id: CLIENT_ID,
  });

  let response: any;
  try {
    response = await fetchFn(tokenEndpoint, {
      method: 'POST',
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      body: body.toString(),
    });
  } catch (err: any) {
    return { success: false, error: `Token refresh request failed: ${err.message}` };
  }

  if (!response.ok) {
    // 400 or 401 means refresh token is invalid/expired
    if (response.status === 400 || response.status === 401) {
      credentials.delete();
      return {
        success: false,
        error: 'Refresh token is invalid or expired. Please run `monocle login --tenant <domain>` to re-authenticate.',
      };
    }
    return { success: false, error: `Token refresh failed (HTTP ${response.status})` };
  }

  const tokenData = await response.json();

  const now = new Date();
  const accessTokenExpiresAt = new Date(now.getTime() + (tokenData.expires_in ?? 3600) * 1000);
  const refreshTokenExpiresAt = new Date(now.getTime() + REFRESH_TOKEN_TTL_DAYS * 24 * 60 * 60 * 1000);

  // Decode ID token to extract email and tenant_name (if new id_token provided)
  let email = currentCredentials.email;
  let tenantName = currentCredentials.tenant_name;
  if (tokenData.id_token) {
    try {
      const payload = decodeIdTokenPayload(tokenData.id_token);
      if (payload.email) email = payload.email;
      if (payload.tenant_name) tenantName = payload.tenant_name;
    } catch {
      // Keep existing values
    }
  }

  const newCredentials: CredentialsData = {
    tenant_domain: currentCredentials.tenant_domain,
    tenant_name: tenantName,
    email: email,
    access_token: tokenData.access_token,
    refresh_token: tokenData.refresh_token ?? currentCredentials.refresh_token,
    id_token: tokenData.id_token ?? currentCredentials.id_token,
    access_token_expires_at: accessTokenExpiresAt.toISOString(),
    refresh_token_expires_at: refreshTokenExpiresAt.toISOString(),
    router_url: discoveredRouterUrl ?? currentCredentials.router_url,
  };

  credentials.write(newCredentials);

  return { success: true, credentials: newCredentials };
}

export function decodeIdTokenPayload(idToken: string): any {
  const parts = idToken.split('.');
  if (parts.length !== 3) {
    throw new Error('Invalid ID token format');
  }
  const payload = Buffer.from(parts[1], 'base64url').toString('utf-8');
  return JSON.parse(payload);
}

File: src/refresh.rs
//! Access-token refresh via the `refresh_token` grant, plus ID-token decoding.

use base64::Engine;
use serde::Deserialize;
use serde_json::Value;

use crate::credentials::{Credentials, CredentialsData};
use crate::error::{AppError, Result};
use crate::net::Client;
use crate::oidc::{discover_oidc, resolve_stark_domain};
use crate::util::{now_ms, to_iso};

const CLIENT_ID: &str = "monocle-cli";
const REFRESH_TOKEN_TTL_DAYS: i64 = 30;

#[derive(Deserialize)]
pub struct TokenResponse {
    pub access_token: Option<String>,
    pub refresh_token: Option<String>,
    pub id_token: Option<String>,
    pub expires_in: Option<i64>,
    #[allow(dead_code)]
    pub token_type: Option<String>,
}

/// Refresh the access token. On success returns the freshly-written credentials;
/// on a 400/401 it deletes the stored credentials (refresh token dead) and
/// returns a re-login message. The error string is what callers print verbatim.
pub fn refresh_access_token(
    client: &Client,
    current: &CredentialsData,
    creds: &Credentials,
) -> Result<CredentialsData> {
    // Discover token endpoint and router URL.
    let (token_endpoint, discovered_router_url) = (|| -> Result<(String, Option<String>)> {
        let stark_domain = resolve_stark_domain(&current.tenant_domain)?;
        let oidc = discover_oidc(client, &stark_domain)?;
        Ok((oidc.token_endpoint, oidc.router_url))
    })()
    .map_err(|e| AppError::new(format!("OIDC Discovery failed: {e}")))?;

    // Request new tokens.
    let resp = client
        .post_form(
            &token_endpoint,
            &[],
            &[
                ("grant_type", "refresh_token"),
                ("refresh_token", &current.refresh_token),
                ("client_id", CLIENT_ID),
            ],
        )
        .map_err(|e| AppError::new(format!("Token refresh request failed: {e}")))?;

    if !resp.ok() {
        if resp.status == 400 || resp.status == 401 {
            creds.delete();
            return Err(AppError::new(
                "Refresh token is invalid or expired. Please run `monocle login --tenant <domain>` to re-authenticate.",
            ));
        }
        return Err(AppError::new(format!(
            "Token refresh failed (HTTP {})",
            resp.status
        )));
    }

    let token: TokenResponse = resp.json()?;

    let now = now_ms();
    let access_expires = to_iso(now + token.expires_in.unwrap_or(3600) * 1000);
    let refresh_expires = to_iso(now + REFRESH_TOKEN_TTL_DAYS * 24 * 60 * 60 * 1000);

    // Decode ID token to refresh email / tenant_name when a new one is provided.
    let mut email = current.email.clone();
    let mut tenant_name = current.tenant_name.clone();
    if let Some(id_token) = &token.id_token {
        if let Ok(payload) = decode_id_token_payload(id_token) {
            if let Some(v) = payload.get("email").and_then(Value::as_str) {
                email = v.to_string();
            }
            if let Some(v) = payload.get("tenant_name").and_then(Value::as_str) {
                tenant_name = v.to_string();
            }
        }
    }

    let new_creds = CredentialsData {
        tenant_domain: current.tenant_domain.clone(),
        tenant_name,
        email,
        access_token: token.access_token.unwrap_or_default(),
        refresh_token: token
            .refresh_token
            .unwrap_or_else(|| current.refresh_token.clone()),
        id_token: token.id_token.unwrap_or_else(|| current.id_token.clone()),
        access_token_expires_at: access_expires,
        refresh_token_expires_at: refresh_expires,
        router_url: discovered_router_url.or_else(|| current.router_url.clone()),
    };

    creds.write(&new_creds)?;
    Ok(new_creds)
}

/// Decode the payload (middle segment) of a JWT into JSON.
pub fn decode_id_token_payload(id_token: &str) -> Result<Value> {
    let parts: Vec<&str> = id_token.split('.').collect();
    if parts.len() != 3 {
        return Err(AppError::new("Invalid ID token format"));
    }
    let bytes = base64::engine::general_purpose::URL_SAFE_NO_PAD
        .decode(parts[1])
        .map_err(|e| AppError::new(e.to_string()))?;
    let value: Value = serde_json::from_slice(&bytes)?;
    Ok(value)
}
